### PR TITLE
Fix timing on iOS Toolbar

### LINKED ISSUE #8335: [iOS] Shell Back Button Label
### Description

When using navigation pages with Shell-based navigation on an iOS .NET MAUI app, the back button label will set itself to the same as the page title if the title is short enough, otherwise it will do the expected behaviours (when page title is longer) and have a label of 'Back' (see images below).

I do not believe that this is intended behaviour as that would not make much sense. I should also add that this does not happen on Windows or on an Android emulator (12.0 - API 31), but I was unable to test the behaviour on MacCatalyst.

### Steps to Reproduce

1. Create a new dotnet MAUI project.
2. Create the MainPage and another page, e.g. DemoPage, and register pages and routes on AppShell.xaml.cs
3. Insert a button with command `Shell.Current.GoToAsync(nameof(DemoPage));` on the MainPage
4. Run the app on an iOS iPhone 15.5 simulator with Hot Reload enabled.
5. Navigate to DemoPage
6. Edit the title parameter in ContentPage there by making it longer and shorter to see the back button's unexpected behaviors on simulator live.

Alternatively clone [this repo](https://github.com/ThatJammyDodger/iOSNav) (https://github.com/ThatJammyDodger/iOSNav) and run it in iOS simulator, using the buttons to see what happens.

**Results:**
Short
Long

Dotnet version: 6.0.400-preview.22301.10

### Version with bug

6.0.400 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Affected platform versions

iOS 15.5

### Did you find any workaround?
https://github.com/PureWeen/ShanedlerSamples/blob/main/ShanedlerSamples/Library/Workarounds/ShellWorkarounds.iOS.cs
Edit: Yes, see comment below
### Relevant log output

_No response_

### LINKED ISSUE #9269: Title view not displayed after navigating shell tabs
### Description

While adding a title view to a new .NET MAUI shell application with a bottom tab bar, I noticed that the title view is removed after returning to first tab. This has only been verified on iOS and macOS.

[Repository containing sample project](https://github.com/tdeborde2/ShellTitleView)

### Steps to Reproduce

1. Create new .NET MAUI project
2. Create new Content View (XAML) and set as shell title view.
3. Add a tab bar and two tabs to the AppShell.xaml.
4. Launch app and navigate to second tab and return to the first tab.

### Version with bug

6.0.486 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, macOS

### Affected platform versions

iOS 15, macOS 12.3.1

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #9687: Shell TitleView disappearing on tab change
### Description

On iOS Simulator with Shell.FlyoutBehavior="Flyout" and Tabs, Shell.TitleView is disappearing on tab change. 

### Steps to Reproduce

1. Create new .Net MAUI project.
2. Add couple ContentPages.
3. Add "AppTitleView" ContentView
3. Edit AppShell.xaml to set Shell.FlyoutBehavior="Flyout"
4. Create couple FlyoutItem with Tab elements.
5. Add 
`
  <Shell.TitleView>
    <shellTitleView:AppTitleView />
  </Shell.TitleView>
`
 to Content Pages
 

6. Run application, switch from "Tab 1" to "Tab 2"
7. Pages on the "Tab 1" will not show "Shell.TitleView", but if you named that TitleView control it still accesable from code behind.
8. Simple project to demostrate the behavior:
[https://github.com/belmonmi/ShellTitleView](url)

### Version with bug

6.0.408

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, I was *not* able test on other platforms

### Affected platform versions

iOS 15.5

### Did you find any workaround?

No

### Relevant log output

```shell
2022-08-26 12:20:38.115611-0500 ShellTitleView[10063:986797] 
Options:
2022-08-26 12:20:38.115788-0500 ShellTitleView[10063:986797]   --bool-flag (Example)
	type: bool  default: false
2022-08-26 12:20:38.115883-0500 ShellTitleView[10063:986797]   --aot-lazy-assembly-load (Load assemblies referenced by AOT images lazily)
	type: bool  default: false
Resolved pending breakpoint for 'ShellTitleView.Program.Main(System.String[])' to C:\Projects\iNet6\Mobile\Test\Issues\ShellTitleView\ShellTitleView\Platforms\iOS\Program.cs:10 [0x00000].
Resolved pending breakpoint for 'Xamarin.HotReload.HotReloadAgent.BreakpointSendToIde(System.String)' to D:\a\_work\1\s\HotReload\Source\Xamarin.HotReload.Agent\HotReloadAgent.cs:419 [0x00000].
Resolved pending breakpoint for 'Xamarin.HotReload.HotReloadAgent.BreakpointCheckpoint()' to D:\a\_work\1\s\HotReload\Source\Xamarin.HotReload.Agent\HotReloadAgent.cs:414 [0x00000].
2022-08-26 12:20:40.312306-0500 ShellTitleView[10063:986797] SecTaskLoadEntitlements failed error=22 cs_flags=200, pid=10063
2022-08-26 12:20:40.312590-0500 ShellTitleView[10063:986797] SecTaskCopyDebugDescription: ShellTitleView[10063]/0#-1 LF=0
2022-08-26 12:20:40.319471-0500 ShellTitleView[10063:986797] SecTaskLoadEntitlements failed error=22 cs_flags=200, pid=10063
2022-08-26 12:20:40.319731-0500 ShellTitleView[10063:986797] SecTaskCopyDebugDescription: ShellTitleView[10063]/0#-1 LF=0
Thread started:  #2
Thread started:  #3
Thread started:  #4
2022-08-26 12:20:41.484090-0500 ShellTitleView[10063:986797] [TableView] Warning once only: UITableView was told to layout its visible cells and other contents without being in the view hierarchy (the table view or one of its superviews has not been added to a window). This may cause bugs by forcing views inside the table view to load and perform layout without accurate information (e.g. table view bounds, trait collection, layout margins, safe area insets, etc), and will also cause unnecessary performance overhead due to extra layout passes. Make a symbolic breakpoint at UITableViewAlertForLayoutOutsideViewHierarchy to catch this in the debugger and see what caused this to occur, so you can avoid this action altogether if possible, or defer it until the table view has been added to a window. Table view: <_UIMoreListTableView: 0x7fa4daef2c00; frame = (0 0; 0 0); clipsToBounds = YES; gestureRecognizers = <NSArray: 0x600002e14330>; layer = <CALayer: 0x600002415140>; contentOffset: {0, 0}; contentSize: {0, 0}; a
djustedContentInset: {0, 0, 0, 0}; dataSource: <UIMoreListController: 0x7fa4d9fa97c0>>
Thread started: .NET Timers #5
Thread started: <Thread Pool> #6
Thread started: .NET ThreadPool Gate #7
Thread started: <Thread Pool> #8
Thread started: <Thread Pool> #9
Thread started: <Thread Pool> #10
Thread started: <Thread Pool> #11
Thread started: <Thread Pool> #12
Thread started: <Thread Pool> #13
Thread started: <Thread Pool> #14
2022-08-26 12:20:45.960535-0500 ShellTitleView[10063:986829] Warning: observer object was not disposed manually with Dispose()
2022-08-26 12:20:45.960790-0500 ShellTitleView[10063:986829] Warning: observer object was not disposed manually with Dispose()
```

### LINKED ISSUE #9842: Shell Title not changing as expected
### Description

The titlebar won't change upon changing between sub items of a flyout item.

### Expected:
App shows title of current page

### Happening:
Title will change to current page if switching from Home to any Test page and back.
Tilte won't change if switching between Test pages.

### Reproduction:
- Start App
- click on `add Shell Item` until you got 2-3 Test pages
- switch between Test pages and ocassionaly to home page

### Link to public reproduction project repository

https://github.com/rxDietel/MauiAppShellTest

### Version with bug

6.0.400

### Last version that worked well

Unknown/Other

### Affected platforms

Windows, iOS, I was *not* able test on other platforms

### Affected platform versions

net6.0-windows10.0.19041.0
iOS 15.6.1

### Did you find any workaround?

I didn't find any workaround for the title bar bug
The Project contains a few more bugs including their workarounds
Everything is described in ReadMe.md

### Relevant log output

_No response_

### LINKED ISSUE #10452: Toolbar items are affected from previous page on Android
### Description

I have a page with three primary items and three secondary items:

image

This page is correct. But when navigating to second page, there should be two primary and two secondaries, but instead there is three primaries:

image

And the third page should have one primary and one secondary items, but both are shown as primary:

image

My guess is that the new page will have always has the same amount as primary items as the previous page. 

This issue is only on Android. It works as expected on Windows at least.

### Steps to Reproduce

1. Create a new MAUI app.
2. Update the main page:

``` XAML
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="MauiIssues.MainPage">

    <ContentPage.ToolbarItems>
        <ToolbarItem Text="Primary 1" />
        <ToolbarItem Text="Primary 2" />
        <ToolbarItem Text="Primary 3" />
        <ToolbarItem Text="Secondary 1"
                     Order="Secondary" />
        <ToolbarItem Text="Secondary 2"
                     Order="Secondary" />
        <ToolbarItem Text="Secondary 3"
                     Order="Secondary" />
    </ContentPage.ToolbarItems>

    <VerticalStackLayout>
        <Label 
            Text="First page. Three primary, and three secondary."
            VerticalOptions="Center" 
            HorizontalOptions="Center" />

        <Button 
            x:Name="ButtonNavigate"
            Text="Go to second page"
            Clicked="ButtonNavigate_Clicked"
            HorizontalOptions="Center" />

    </VerticalStackLayout>
</ContentPage>
```

``` CS
public partial class MainPage : ContentPage
{
	public MainPage()
	{
		InitializeComponent();
	}

	private async void ButtonNavigate_Clicked(object sender, EventArgs e)
	{
        await Shell.Current.GoToAsync(nameof(SecondaryPage));
	}
}
```

3. Add second page:

``` XAML
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="MauiIssues.SecondaryPage"
             Title="Second">
    <ContentPage.ToolbarItems>
        <ToolbarItem Text="Primary 1" />
        <ToolbarItem Text="Primary 2" />
        <ToolbarItem Text="Secondary 1"
                     Order="Secondary" />
        <ToolbarItem Text="Secondary 2"
                     Order="Secondary" />
    </ContentPage.ToolbarItems>
    
    <VerticalStackLayout>
        <Label 
            Text="Second page. Two primary, and two secondary."
            VerticalOptions="Center" 
            HorizontalOptions="Center" />

        <Button 
            x:Name="ButtonNavigate"
            Text="Go to third page"
            Clicked="ButtonNavigate_Clicked"
            HorizontalOptions="Center" />

    </VerticalStackLayout>
</ContentPage>
``` 

``` CS
public partial class SecondaryPage : ContentPage
{
	public SecondaryPage()
	{
		InitializeComponent();
	}

	private async void ButtonNavigate_Clicked(object sender, EventArgs e)
	{
		await Shell.Current.GoToAsync(nameof(ThirdPage));
	}
}
``` 

4. Add third page:

``` XAML
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="MauiIssues.ThirdPage"
             Title="Third">
    <ContentPage.ToolbarItems>
        <ToolbarItem Text="Primary 1" />
        <ToolbarItem Text="Secondary 1"
                     Order="Secondary" />
    </ContentPage.ToolbarItems>

    <VerticalStackLayout>
        <Label 
            Text="Third page. One primary, and one secondary."
            VerticalOptions="Center" 
            HorizontalOptions="Center" />
    </VerticalStackLayout>
</ContentPage>
```

``` CS
public partial class ThirdPage : ContentPage
{
	public ThirdPage()
	{
		InitializeComponent();
	}
}
```

### Link to public reproduction project repository

https://github.com/pekspro/MauiIssues/tree/10452_Toolbar_items_are_affected_from_previous_page

### Version with bug

6.0.486 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 11

### Did you find any workaround?

Adding the code below to an affected page does solve the problem. But there will be some flickering on navigation.

``` CS
#if ANDROID
    protected override void OnNavigatedTo(NavigatedToEventArgs args)
    {
        base.OnNavigatedTo(args);

        var secondayToolBarItems = ToolbarItems.Where(x => x.Order == ToolbarItemOrder.Secondary).ToList();

        foreach (var item in secondayToolBarItems)
        {
            ToolbarItems.Remove(item);
        }

        Application.Current.Dispatcher.Dispatch(() =>
        {
            foreach (var item in secondayToolBarItems)
            {
                ToolbarItems.Add(item);
            }
        });
    }
#endif

```

### Relevant log output

_No response_

### LINKED ISSUE #10945: [iOS] Title is bugged during and after page transition
### Description

The following bugs are present:
- The new title only shows up once the animation of page transition is over
- Title gets changed for all the pages in a stack to whatever the current title is supposed to be, so in the screenshots bellow MainPage title becomes "sub"

Screenshot 2022-10-26 at 22 26 37
Screenshot 2022-10-26 at 22 26 52

### Steps to Reproduce

Just navigate to a child page in app shell on iOS, sample repo bellow is available which shows the bug.

### Link to public reproduction project repository

https://github.com/harry-cpp/MauiBugSamples/tree/main/iOSBuggedTitle

### Version with bug

6.0.486 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Affected platform versions

All

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #11691: iOS Navigation / Back Button has page title of current page not previous page; slow/delayed to change navigation page
### Description

When using Shell.Current.GoToAsync to navigate to a child page, the back button presented in iOS would have a left arrow / reverse disclosure indicator, and the previous page's title if it can fit. However, the title presented when on the child page is of the same page you are on, not the page you were just on previously. This can be confusing as users expect the text next to the back button to show where they were last. 

Also, the titles lag and show the previous page's title for a few moments after navigation. Xamarin Forms on iOS had a smooth iOS navigation transition where it sort of slid in. I think that's a native behavior that MAUI has lost. 

https://user-images.githubusercontent.com/3627193/204161163-097dc832-1eb2-46ae-aae8-13911f7313ce.mp4

### Steps to Reproduce

1. Create a File -> New Maui app
2. Create a page that you can navigate to that should produce a back button in the Navigation Bar
3. Give both pages a title
4.  Navigate to that new page. The title displayed next to the back button should be the parent page's title, not the current page's title

### Link to public reproduction project repository

https://github.com/edgiardina/MauiBug_WrongBackButtonTitle

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Affected platform versions

iOS 16

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change

With MAUI we moved the code for processing `Toolbar` logic up into the xplat layer by way of an xplat `Toolbar` control. This allowed us to consolidate the toolbar logic that was copy and pasted across all platforms and between shell and non-shell. On iOS the shell toolbar management is handled by `ShellPageToolbarTracker` which gets instantiated for every single page. The code inside that tracker needs to ignore any `toolbar` updates if it's not the currently visible page. 

iOS is also a little bit tricky to wire up with regards to timing the back button title text. The BackButton text is based on the  `UIViewController.NavigationItem` of the previous page so we have to pull the `BackButtonBehavior` off the incoming page and set it on the `NavItem` of the previous page before the current page is visible.  `NavigationPage` gets around this by making it so the user has to set `NavigationPage.BackButtonTitle` on the previous page. `NavigationPage` basically replicates how `iOS` propagates back button text where as `Shell` lets you set the back button text on the currently visible page. 

### Issues Fixed

Fixes #10945
Fixes #11691
Fixes #9269
Fixes #9842
Fixes #9687
Fixes #8335 
Fixed #10452

### CODE REVIEW

**Comment by PureWeen** [comment]
@rmarinho I tested this on a 13.7 and a 14.5 iPhone

Let me know if I should test back further from that.

File: src/TestUtils/src/DeviceTests/AssertionExtensions.iOS.cs [hunk 1]
```diff
@@ -546,5 +547,56 @@ public static UIView GetAccessiblePlatformView(this UIView platformView)
...
+		{
+			var item = uINavigationBar.FindDescendantView<UIView>(result =>
+			{
+				return result.Class.Name?.Contains("UINavigationBarTitleControl", StringComparison.OrdinalIgnoreCase) == true;
+			});
+
+			//Pre iOS 15
+			item = item ?? uINavigationBar.FindDescendantView<UIView>(result =>
+			{
+				return result.Class.Name?.Contains("UINavigationBarContentView", StringComparison.OrdinalIgnoreCase) == true;
```
Comment (PureWeen) [inline src/TestUtils/src/DeviceTests/AssertionExtensions.iOS.cs:581]: @rmarinho

**Review by rmarinho — CHANGES_REQUESTED** [review]
@PureWeen  seems is failing just 1 test now 

https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=7243580&view=ms.vss-test-web.build-test-results-tab

**Review by rmarinho — APPROVED** [review]

**Comment by PureWeen** [comment]
Adding this one back into the pool for some discussion. This one fixes a bunch of issues and I think we can break apart a few parts of this one. 

For example, I think just backporting the xplat changes [here](https://github.com/dotnet/maui/pull/12888/files#diff-953760fc28b84326460f63e7ca19e18a6554962c946d851a456948b1b5db3e85R1509-R1516) will fix a number of issues

**Comment by angelru** [comment]
When will these changes be available?

**Pull request merged**